### PR TITLE
Support X-Last-Seen-Txn.

### PR DESCRIPTION
**Pull request by bitbckt**
Once again, I don't entirely know what I'm doing here.

### CODE REVIEW

File: FaunaDB.Client/Client/FaunaClient.cs [hunk 1]
```diff
@@ -122,6 +122,13 @@ public FaunaClient(IClientIO clientIO)
 
             RaiseForStatusCode(responseHttp);
 
+            if (responseHttp.ResponseHeaders.ContainsKey("X-Txn-Time")) {
+                // there shouldn't ever be more than one...
```
Comment (marrony) [inline FaunaDB.Client/Client/FaunaClient.cs]: what?

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 2]
```diff
@@ -49,6 +49,11 @@ public DefaultClientIO(string secret, Uri endpoint, TimeSpan timeout)
             message.Content = dataString;
             message.Headers.Authorization = authHeader;
 
+            var lastSeen = LastSeen.Txn;
```
Comment (marrony) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: what if you make `LastSeen` a class variable....

File: FaunaDB.Client/Client/LastSeen.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
+using System;
+using System.Threading;
+
+/// <summary>
+///   A thread-safe static helper used to store the last seen
+///   transaction time returned by the database.
+/// </summary>
+namespace FaunaDB.Client {
+    public class LastSeen {
+        private static long lastSeenTxn = 0;
```
Comment (marrony) [inline FaunaDB.Client/Client/LastSeen.cs]: and remove the static qualifier?

File: FaunaDB.Client/Client/FaunaClient.cs (hunk 1 above)
Comment (bitbckt) [reply FaunaDB.Client/Client/FaunaClient.cs]: More than one X-Txn-Time. Too vague?

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 2 above)
Comment (bitbckt) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: That would share per client, but not across clients. The expected concurrency in these clients isn't clear to me from one to another, so I decided to make this global. Is per-client "better"?

File: FaunaDB.Client/Client/FaunaClient.cs (hunk 1 above)
Comment (marrony) [reply FaunaDB.Client/Client/FaunaClient.cs]: doesn't `responseHttp.ResponseHeaders` represent all the response headers?

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 2 above)
Comment (bitbckt) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: (I think I have that right... C# is also not at all my thing)

Comment (marrony) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: well, given this is a library, make things global might make the users code unpredictable

File: FaunaDB.Client/Client/FaunaClient.cs (hunk 1 above)
Comment (bitbckt) [reply FaunaDB.Client/Client/FaunaClient.cs]: Hmm. I interpreted the type to be a dictionary with string -> enum[string]...

Comment (bitbckt) [reply FaunaDB.Client/Client/FaunaClient.cs]: .NET apparently supports multiple-value headers? Weird. https://msdn.microsoft.com/en-us/library/system.net.http.headers.httpheaders(v=vs.118).aspx

File: FaunaDB.Client/Client/DefaultClientIO.cs [hunk 4]
```diff
@@ -20,22 +20,28 @@ class DefaultClientIO : IClientIO
...
         { }
 
         public IClientIO NewSessionClient(string secret) =>
-            new DefaultClientIO(client, AuthHeader(secret));
+            new DefaultClientIO(client, AuthHeader(secret), lastSeen);
 
         public Task<RequestResult> DoRequest(HttpMethodKind method, string path, string data, IReadOnlyDictionary<string, string> query = null) =>
             DoRequestAsync(method, path, data, query);
 
+        public void UpdateLastSeen(long time) =>
```
Comment (erickpintor) [inline FaunaDB.Client/Client/DefaultClientIO.cs]: Can this be private or internal?

File: FaunaDB.Client/Client/LastSeen.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                initial = lastSeenTxn;
+
+                // don't move the last seen time backward
+                if (time <= initial) {
+                    return;
+                }
+            } while (initial != Interlocked.CompareExchange(ref lastSeenTxn, time, initial));
+
+            return;
```
Comment (erickpintor) [inline FaunaDB.Client/Client/LastSeen.cs]: Do you need this return?

Comment (bitbckt) [reply FaunaDB.Client/Client/LastSeen.cs]: Apparently not. Java-ism, I guess.

File: FaunaDB.Client/Client/DefaultClientIO.cs (hunk 4 above)
Comment (bitbckt) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: It's an interface method, called from the client - I think it needs to be public?

Comment (marrony) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: you don't need to expose this method on the interface, you can do all the job on `DefaultClientIO`

Comment (bitbckt) [reply FaunaDB.Client/Client/DefaultClientIO.cs]: Aha. So I can. Coming right up.

**Comment by bitbckt** [comment]
I'm using `csharp-mode`. ;-)

**Pull request merged**